### PR TITLE
Feature/cdap 6168 ds read write

### PR DESCRIPTION
**Pull request by chtyim**
This PR seems big, but the core changes are actually in the following classes:
- Changes in cdap-api for new annotations
- `DatasetRuntimeContext`
- `DefaultDatasetRuntimeContext`
- `InterceptableClassLoader`, `MainClassLoader` and `ProgramClassLoader`
- `DatasetClassRewriter`
- `LineageWriterDatasetFramework`
- Annotating existing dataset classes.

This PR has two commits. Suggested to review them separate.
1. (CDAP-6168) Add access control to Dataset
   - Added new annotations to API 
     - @NoAccess, @ReadOnly, @WriteOnly, @ReadWrite
   - Modify program classloader to intercept custom dataset classes methods
     - Analyze method annotation
     - Call DatasetRuntimeContext.onMethodEnter and DatasetRuntimeContext.onMethodExit
     - Perform authorization, lineage and usage tracking (ALU) per individual dataset operation
   - Modify main classloader to intercept system dataset classes methods
     - Same as kind of interception as custom dataset classes
   - Modify runtime to use the new main classloader
     - Unit-test framework, introduced a TestRunner
     - Standalone
     - Program TwillRunnable
     - Spark driver and executor containers
     - MapReduce containers
     - Not modifying the Master since it shouldn't perform dataset operation in user namespace
       - We don't do ALU for system datasets
   - Fixed Spark logging context and stdout/stderr redirection
     - Not really related to CDAP-6168, but fix it anywhere as the change is small
2. (CDAP-6168) Added annotations to system datasets

### CODE REVIEW

**Comment by chtyim** [comment]
More authorization unit-test in https://issues.cask.co/browse/CDAP-6648

**Comment by chtyim** [comment]
Building: http://builds.cask.co/browse/CDAP-DUT4428-23

File: cdap-api/src/main/java/co/cask/cdap/api/dataset/lib/IndexedObjectStore.java [hunk 1]
```diff
@@ -140,6 +145,7 @@ public T read(String key) {
    * @param object object to be stored
    * @param secondaryKeys indices that can be used to lookup the object
    */
+  @WriteOnly
```
Comment (anew) [inline cdap-api/src/main/java/co/cask/cdap/api/dataset/lib/IndexedObjectStore.java:148]: Interesting, this seems like a write-only, but in order to maintain the index it actually involves a read (of the old value). So will the enforcement fail then? Or does this not read the old values of secondary keys?

File: cdap-api/src/main/java/co/cask/cdap/api/dataset/lib/IndexedTable.java [hunk 2]
```diff
@@ -247,6 +258,7 @@ private void assertIndexedColumn(byte[] column) {
    * 
    * @param put The put operation to store
    */
+  @ReadWrite
```
Comment (anew) [inline cdap-api/src/main/java/co/cask/cdap/api/dataset/lib/IndexedTable.java]: IndexedObjectStore.write is @WriteOnly, but this is @ReadWrite. Why?

File: cdap-api/src/main/java/co/cask/cdap/api/dataset/lib/IndexedTable.java [hunk 3]
```diff
@@ -542,50 +567,59 @@ public void increment(Increment increment) {
...
 
+  @ReadOnly
   @Override
   public SplitReader<byte[], Row> createSplitReader(Split split) {
     return table.createSplitReader(split);
   }
 
   /* BatchWritable implementation */
 
+  @WriteOnly
```
Comment (anew) [inline cdap-api/src/main/java/co/cask/cdap/api/dataset/lib/IndexedTable.java:618]: put() is @ReadWrite and it is called from here. Why is this @WriteOnly then?

File: cdap-api/src/main/java/co/cask/cdap/internal/dataset/DatasetRuntimeContext.java [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+  }
+
+  /**
+   * Sets the current {@link DatasetRuntimeContext}. This method can only be initiated from system class.
+   */
+  public static Cancellable setContext(DatasetRuntimeContext context) {
+    final Class[] callerClasses = CallerClassSecurityManager.getCallerClasses();
+    // 0 is the CallerClassSecurityManager, 1 is this class, hence 2 is the actual caller
+    if (callerClasses.length < 3) {
+      throw new IllegalStateException("Invalid call stack.");
```
Comment (anew) [inline cdap-api/src/main/java/co/cask/cdap/internal/dataset/DatasetRuntimeContext.java:69]: add to the exception what the actual caller classes are?

File: cdap-api/src/main/java/co/cask/cdap/internal/dataset/DatasetRuntimeContext.java [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+    }
+    // This is the guard against if someone tries to call this method outside of CDAP system (e.g. from user code)
+    if (callerClasses[2].getClassLoader() != DatasetRuntimeContext.class.getClassLoader()) {
+      throw new IllegalAccessError("Not allow to set context from " + callerClasses[2]);
+    }
```
Comment (anew) [inline cdap-api/src/main/java/co/cask/cdap/internal/dataset/DatasetRuntimeContext.java:75]: if someone was able to bypass this, what would be the impact on security? Say, user code calls DatasetRuntimeContext.class.getClassLoader() inside app code. Then he can instantiate any class from that class loader and thus call this method. Would he be able to bypass the access enforcement?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactRepository.java [hunk 6]
```diff
@@ -592,7 +592,6 @@ private ArtifactClasses inspectArtifact(Id.Artifact artifactId, File artifactFil
     } else {
       return ArtifactClasses.builder()
         .addApps(artifactClasses.getApps())
-        .addDatasets(artifactClasses.getDatasets())
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactRepository.java:595]: Does this PR reverts a lot of the changes made in the previous PRs for this feature? (6012 and 6038)? or maybe it just seems so to me...

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/distributed/ContainerLauncherGenerator.java [hunk 7]
```diff
@@ -74,14 +75,16 @@ public static void generateLauncherJar(String launcherClassPath, String classLoa
   /**
    * Generates a JAR file that contains a class with a static main method.
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/distributed/ContainerLauncherGenerator.java]: contains zero or more classes with a main method?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/AbstractProgramTwillRunnable.java [hunk 8]
```diff
@@ -371,7 +371,6 @@ private ProgramOptions createProgramOptions(CommandLine cmdLine, TwillContext co
     Map<String, String> arguments = Maps.newHashMap(original.getArguments().asMap());
     arguments.put(ProgramOptionConstants.INSTANCE_ID, Integer.toString(context.getInstanceId()));
     arguments.put(ProgramOptionConstants.INSTANCES, Integer.toString(context.getInstanceCount()));
-    arguments.put(ProgramOptionConstants.RUN_ID, original.getArguments().getOption(ProgramOptionConstants.RUN_ID));
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/AbstractProgramTwillRunnable.java:374]: why runId is not needed any longer?

File: cdap-common/src/main/java/co/cask/cdap/common/app/MainClassLoader.java [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+
+  /**
+   * Returns the list of {@link URL} for the system ClassLoader.
+   */
+  private static URL[] getClassPath() {
+    List<URL> urls = new ArrayList<>();
+    ClassLoader sysClassLoader = ClassLoader.getSystemClassLoader();
+
+    if (sysClassLoader instanceof URLClassLoader) {
+      urls.addAll(Arrays.asList(((URLClassLoader) sysClassLoader).getURLs()));
```
Comment (anew) [inline cdap-common/src/main/java/co/cask/cdap/common/app/MainClassLoader.java]: return ((URLClassLoader) sysClassLoader).getURLs()

Or is there a reason for copying a list and then into a new array?

File: cdap-common/src/main/java/co/cask/cdap/common/app/MainClassLoader.java [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+    }
+  }
+
+  @Override
+  public byte[] rewriteClass(String className, InputStream input) throws IOException {
+    return datasetRewriter.rewriteClass(className, input);
+  }
+
+  /**
+   * Returns the list of {@link URL} for the system ClassLoader.
```
Comment (anew) [inline cdap-common/src/main/java/co/cask/cdap/common/app/MainClassLoader.java]: array, not list

File: cdap-common/src/main/java/co/cask/cdap/common/dataset/DatasetClassRewriter.java [hunk 11]
```diff
@@ -0,0 +1,251 @@
...
+    // Collects the set of methods that doesn't need rewriting
+    Set<Method> ignoreMethods = new HashSet<>();
+
+    // Should ignore all methods in Object.
+    for (java.lang.reflect.Method method : Object.class.getDeclaredMethods()) {
+      ignoreMethods.add(Method.getMethod(method));
+    }
+
+    // Ignore methods in interface that Dataset implements, which are not part of dataset operation
+    for (Class<?> cls : Arrays.asList(TransactionAware.class, Closeable.class, MeteredDataset.class)) {
```
Comment (anew) [inline cdap-common/src/main/java/co/cask/cdap/common/dataset/DatasetClassRewriter.java]: what about InputFormatProvider? OutputCommitterProvider?

File: cdap-common/src/main/java/co/cask/cdap/common/dataset/DatasetClassRewriter.java [hunk 12]
```diff
@@ -0,0 +1,251 @@
...
+          if (isConstructor) {
+            // this._datasetRuntimeContext = DatasetRuntimeContext.getContext();
+            loadThis();
+            invokeStatic(DATASET_RUNTIME_CONTEXT_TYPE,
+                         new Method("getContext", Type.getMethodDescriptor(DATASET_RUNTIME_CONTEXT_TYPE)));
+            putField(datasetType, datasetRuntimeContextField, DATASET_RUNTIME_CONTEXT_TYPE);
+          }
+
+          Type methodAnnotationType = getMethodAnnotationType(hasRead, hasWrite, hasNoAccess);
+          Type defaultAnnotationType = isConstructor ? Type.getType(NoAccess.class) : Type.getType(ReadWrite.class);
```
Comment (anew) [inline cdap-common/src/main/java/co/cask/cdap/common/dataset/DatasetClassRewriter.java]: would it make sense to have an UnknownAccess as the default if neither of the other ones is present? It would be enforced as if it is a @ReadWrite, but for lineage or audit purpose, it could be logged as Unkown access.

File: cdap-common/src/main/java/co/cask/cdap/common/io/ForwardingLocation.java [hunk 13]
```diff
@@ -0,0 +1,139 @@
...
+  }
+
+  @Override
+  public LocationFactory getLocationFactory() {
+    return delegate.getLocationFactory();
+  }
+
+  @Override
+  public String toString() {
+    return toURI().toString();
```
Comment (anew) [inline cdap-common/src/main/java/co/cask/cdap/common/io/ForwardingLocation.java]: why not delegate.toString()?

File: cdap-common/src/main/java/co/cask/cdap/common/lang/ClassRewriter.java [hunk 14]
```diff
@@ -14,19 +14,15 @@
  * the License.
  */
 
-package co.cask.cdap.app.customds;
```
Comment (anew) [inline cdap-common/src/main/java/co/cask/cdap/common/lang/ClassRewriter.java:17]: Probably a funny git glitch, to treat this as a rename :)

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/DefaultDatasetRuntimeContext.java [hunk 15]
```diff
@@ -0,0 +1,222 @@
...
+      }
+
+      @Override
+      protected Deque<Class<? extends Annotation>> childValue(Deque<Class<? extends Annotation>> parentValue) {
+        // Copy the top of the stack
+        Deque<Class<? extends Annotation>> stack = new ArrayDeque<>(10);
+        Class<? extends Annotation> annotation = parentValue.peekLast();
+
+        // In normal case it shouldn't be null. However, in case the dataset calls the onMethodExit
+        // intentionally, then this can be null.
```
Comment (anew) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/DefaultDatasetRuntimeContext.java]: should we allow that? Wouldn't that be an attempt to bypass access auth?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/DefaultDatasetRuntimeContext.java [hunk 16]
```diff
@@ -0,0 +1,222 @@
...
+    if (annotation == null) {
+      // If there is no annotation on the method, either use immediate parent one from the call stack
+      // or use the default one if the call stack is empty
+      annotation = (parentAnnotation == null) ? defaultAnnotation : parentAnnotation;
+    }
+
+    // Do a quick check if the current permission is allowed by the immediate parent if there is one.
+    if (parentAnnotation != null && !isAllowed(parentAnnotation, annotation)) {
+      throw new DataSetException(
+        String.format("Current scope only allows @%s, but the current method requires @%s",
```
Comment (anew) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/DefaultDatasetRuntimeContext.java]: This logic seems correct. But: IndexedObjectStore.write() is annotated with @WriteOnly, and it calls Table.get() which is implemented as a @ReadOnly method in BufferedTable. So that should trigger this exception. Either we don't have test cases for IndexedObjectStore, or there is a bug in the construction of the call stack.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/DefaultDatasetRuntimeContext.java [hunk 17]
```diff
@@ -0,0 +1,222 @@
...
+    // Record the access (lineage, audit).
+    recordAccess(annotation);
+
+    callStack.addLast(annotation);
+  }
+
+  @Override
+  public void onMethodExit() {
+    // The cancellable should be called when the method exit, which should happen in the same thread
+    // as the method entry call.
```
Comment (anew) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/DefaultDatasetRuntimeContext.java:165]: if onMethodEntry() throws an exception, will this still be called? In that case, we would remove the last entry in the call stack, even though onMethodEntry() never added it.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/cube/CubeDataset.java [hunk 18]
```diff
@@ -62,36 +63,43 @@ public CubeDataset(String name, MetricsTable entityTable,
...
     cube.add(facts);
   }
 
+  @ReadOnly
   @Override
   public Collection<TimeSeries> query(CubeQuery query) {
     return cube.query(query);
   }
 
+  @ReadOnly
```
Comment (anew) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/cube/CubeDataset.java]: This should be @WriteOnly

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/cube/DefaultCube.java [hunk 19]
```diff
@@ -294,6 +303,7 @@ public void delete(CubeDeleteQuery query) {
    * Sets {@link MetricsCollector} for metrics reporting.
    * @param metrics {@link MetricsCollector} to set.
    */
+  @Override
```
Comment (anew) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/cube/DefaultCube.java:298]: does this have to be @NoAccess?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/external/ExternalDataset.java [hunk 20]
```diff
@@ -45,21 +48,25 @@ public ExternalDataset(Map<String, String> runtimeArgs) {
       this.outputFormatClassName != null ? runtimeArgs : Collections.<String, String>emptyMap();
   }
 
+  @ReadOnly
```
Comment (anew) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/external/ExternalDataset.java]: why are these methods (this and the following 3) not @NoAccess? They simply query configuration, but they don't perform actual data access.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/CoreDatasetsModule.java [hunk 21]
```diff
@@ -67,7 +66,6 @@ public void register(DatasetDefinitionRegistry registry) {
     registry.add(new CounterTimeseriesTableDefinition(CounterTimeseriesTable.class.getName(), tableDef));
 
     // in-memory table
-    registry.add(new InMemoryTableDefinition(MemoryTable.class.getName()));
```
Comment (anew) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/CoreDatasetsModule.java:70]: why remove this?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/writer/LineageWriterDatasetFramework.java [hunk 22]
```diff
@@ -101,65 +122,50 @@ public void deleteAllInstances(Id.Namespace namespaceId) throws DatasetManagemen
...
+          public T call() throws Exception {
+            return LineageWriterDatasetFramework.super.getDataset(datasetInstanceId, arguments, classLoader,
+                                                                  classLoaderProvider, owners, accessType);
+          }
+        });
+    } catch (IOException | DatasetManagementException e) {
+      throw e;
     } catch (Exception e) {
-      LOG.warn("Failed to register usage of {} -> {}", owners, datasetInstanceId, e);
+      throw new DatasetManagementException("Failed to craete dataset instance: " + datasetInstanceId, e);
```
Comment (anew) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/writer/LineageWriterDatasetFramework.java]: The new message does not make sense to me. The old one seems more accurate.

File: cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkRuntimeContextProvider.java [hunk 23]
```diff
@@ -183,6 +184,7 @@ public void run() {
         injector.getInstance(SecureStore.class),
         injector.getInstance(SecureStoreManager.class)
       );
+      LoggingContextAccessor.setLoggingContext(sparkRuntimeContext.getLoggingContext());
```
Comment (anew) [inline cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkRuntimeContextProvider.java:187]: unrelated change?

**Comment by anew** [comment]
@chtyim I did  pass over this and left some comments.

Also a question on the approach: We now require that every dataset method (also for custom datasets) is annotated with an access type. Otherwise it will be treated as Read+Write. Then we inject code that validates, as this dataset's methods make calls to its embedded datasets (or to other methods), that these methods are annotated with an access type that is "weaker" than the caller's access type. 

What if instead, for a method that are not annotated, we simply derive the access types from the access types of the methods that it calls. That would a lot easier for a developer, because the correct access type will always be generated auto-magically. It will also avoid the irritation caused by auth errors, or lineage/audit records that assume ReadWrite because a method was not annotated, even though that method may not perform any dataset access. (Like a getter for a config parameter).

**Comment by chtyim** [comment]
I had similar thoughts in the pass couple days. The current design is too restrictive and is hard to annotate it right at every single case. I will revise the approach and update the design. The code change shouldn't be big as the framework is pretty much there.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactRepository.java (hunk 6 above)
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactRepository.java:595]: It does. With the new ClassLoader, actually we don't need to grab the dataset classes during artifact deployment. Whether a class is a dataset class is determined dynamically during classloading.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/AbstractProgramTwillRunnable.java (hunk 8 above)
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/AbstractProgramTwillRunnable.java:374]: It's already inside the `original.getArguments()` (line 371 above), hence not need to set it again.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/CoreDatasetsModule.java (hunk 21 above)
Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/CoreDatasetsModule.java:70]: There is no class implementing `MemoryTable` interface, hence this registration is useless.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/writer/LineageWriterDatasetFramework.java (hunk 22 above)
Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/writer/LineageWriterDatasetFramework.java]: The usage registration won't throw exception in the try block above. Usage registration is now done during dataset constructor/method call, and the exception is handling inside those call (failure in usage recording won't cause failure of the operation).

File: cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkRuntimeContextProvider.java (hunk 23 above)
Comment (chtyim) [inline cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkRuntimeContextProvider.java:187]: It is unrelated, but fixes the Spark logging issue (I was debugging with logs and found this issue).

**Comment by chtyim** [comment]
Rebased and resolved conflicts with release/3.5 branch.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/DefaultDatasetRuntimeContext.java (hunk 17 above)
Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/DefaultDatasetRuntimeContext.java:165]: no it won't be called. The generated code looks like this:

``` java
void method() {
  // the onMethodEntry() will throw exception if the user is not allowed to access this method
  onMethodEntry();
  try {
  } finally {
    onMethodExit();
  }
}
```

File: cdap-api/src/main/java/co/cask/cdap/api/dataset/lib/IndexedObjectStore.java (hunk 1 above)
Comment (chtyim) [inline cdap-api/src/main/java/co/cask/cdap/api/dataset/lib/IndexedObjectStore.java:148]: With the new lineage handling, it will record a `WRITE` lineage

File: cdap-api/src/main/java/co/cask/cdap/api/dataset/lib/IndexedTable.java (hunk 3 above)
Comment (chtyim) [inline cdap-api/src/main/java/co/cask/cdap/api/dataset/lib/IndexedTable.java:618]: With the new lineage handling, a `WRITE` lineage will be record, while a READ and WRITE authorization will be enforced and audit log being published when the actual data operations happen

File: cdap-api/src/main/java/co/cask/cdap/internal/dataset/DatasetRuntimeContext.java (hunk 4 above)
Comment (chtyim) [inline cdap-api/src/main/java/co/cask/cdap/internal/dataset/DatasetRuntimeContext.java:69]: This condition should never happen as there should be someone calling this method. Added a comment.

File: cdap-api/src/main/java/co/cask/cdap/internal/dataset/DatasetRuntimeContext.java (hunk 5 above)
Comment (chtyim) [inline cdap-api/src/main/java/co/cask/cdap/internal/dataset/DatasetRuntimeContext.java:75]: The user needs to be able to define a custom class through the CDAP system classloader (i.e. the one from DatasetRuntimeContext.class.getClassLoader()). With the current `SecurityManager` setting, it is not entirely impossible, though is not easy. I added a check of the class name as well to make sure it comes from the `co.cask.cdap.data2.dataset2.DefaultDatasetRuntimeContext`.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/distributed/ContainerLauncherGenerator.java (hunk 7 above)
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/distributed/ContainerLauncherGenerator.java]: fixed

File: cdap-common/src/main/java/co/cask/cdap/common/app/MainClassLoader.java (hunk 9 above)
Comment (chtyim) [inline cdap-common/src/main/java/co/cask/cdap/common/app/MainClassLoader.java]: Actually this `if` is unnecessary since we already check in the `createFromContext` method that if the classloader is an instance of `URLClassLoader`, hence this condition will never happen. Removed the if.

File: cdap-common/src/main/java/co/cask/cdap/common/app/MainClassLoader.java (hunk 10 above)
Comment (chtyim) [inline cdap-common/src/main/java/co/cask/cdap/common/app/MainClassLoader.java]: fixed

File: cdap-common/src/main/java/co/cask/cdap/common/dataset/DatasetClassRewriter.java (hunk 11 above)
Comment (chtyim) [inline cdap-common/src/main/java/co/cask/cdap/common/dataset/DatasetClassRewriter.java]: This exclusion is no longer needed with the new design.

File: cdap-common/src/main/java/co/cask/cdap/common/io/ForwardingLocation.java (hunk 13 above)
Comment (chtyim) [inline cdap-common/src/main/java/co/cask/cdap/common/io/ForwardingLocation.java]: fixed.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/cube/DefaultCube.java (hunk 19 above)
Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/cube/DefaultCube.java:298]: No longer need the `@NoAccess` annotation with the new logic.

**Comment by anew** [comment]
Hmmm. After the rebase I am not sure what I have to (re-)review. All 3.5K lines? I am going to look at the last two commits, please let me know if anything changed in the first two.

**Comment by chtyim** [comment]
The first commit doesn't change. I overwrite the 2nd one and the 3rd and 4th are new commits.

File: cdap-common/src/main/java/co/cask/cdap/common/dataset/DatasetClassRewriter.java [hunk 24]
```diff
@@ -225,7 +216,7 @@ private Type getMethodAnnotationType(boolean hasRead, boolean hasWrite, boolean
           if (hasWrite) {
             return Type.getType(WriteOnly.class);
           }
-          return Type.getType(NoAccess.class);
+          return null;
```
Comment (anew) [inline cdap-common/src/main/java/co/cask/cdap/common/dataset/DatasetClassRewriter.java:196]: null or UNKNOWN?

Comment (chtyim) [inline cdap-common/src/main/java/co/cask/cdap/common/dataset/DatasetClassRewriter.java:196]: We don't have an unknown annotation class, hence using `null` to represent an un-annotated method.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/DefaultDatasetRuntimeContext.java [hunk 25]
```diff
@@ -130,93 +124,123 @@ private DefaultDatasetRuntimeContext(AuthorizationEnforcer enforcer, DatasetAcce
...
+        throw new DataSetException("Unsupported annotation " + annotation + " on dataset " + datasetId);
+      }
 
-    // Do a quick check if the current permission is allowed by the immediate parent if there is one.
-    if (parentAnnotation != null && !isAllowed(parentAnnotation, annotation)) {
-      throw new DataSetException(
-        String.format("Current scope only allows @%s, but the current method requires @%s",
-                      parentAnnotation.getSimpleName(), annotation.getSimpleName()));
-    }
+      // Perform authorization if the method is annotated.
```
Comment (anew) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/DefaultDatasetRuntimeContext.java:145]: Does the enforcer cache what it has enforced already? Seems that it does not necessarily do... Can we do this enforcement only if the access type diverges from what the parent declared?

Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/DefaultDatasetRuntimeContext.java:145]: The `enforcer.enforce` was written in a very efficient way and expected to get called many times. I think it's better to always call enforce in case if user is trying to manipulate the callstack.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/writer/LineageWriterDatasetFramework.java [hunk 26]
```diff
@@ -101,91 +127,118 @@ public void deleteAllInstances(Id.Namespace namespaceId) throws DatasetManagemen
...
+          @Override
+          public T call() throws Exception {
+            return LineageWriterDatasetFramework.super.getDataset(datasetInstanceId, arguments, classLoader,
+                                                                  classLoaderProvider, owners, accessType);
+          }
+        });
+    } catch (IOException | DatasetManagementException e) {
+      throw e;
     } catch (Exception e) {
-      LOG.warn("Failed to register usage of {} -> {}", owners, datasetInstanceId, e);
```
Comment (anwar6953) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/writer/LineageWriterDatasetFramework.java:162]: nit - typo - "craete" -> "create"

**Comment by anwar6953** [comment]
btw - flow wasn't working, due to stack trace as below.

Following PR has functional changes that get it to work: https://github.com/caskdata/cdap/pull/6292
Note that these changes were only applied to the rc branch, and not the release/3.5 branch.

```
Caused by: java.lang.NullPointerException: null
        at co.cask.cdap.data2.audit.AuditPublishers.publishAccess(AuditPublishers.java:90) ~[cdap-data-fabric-3.5.0.jar:na]
        at co.cask.cdap.data2.metadata.writer.LineageWriterDatasetFramework.publishAudit(LineageWriterDatasetFramework.java:200) ~[cdap-data-fabric-3.5.0.jar:na]
        at co.cask.cdap.data2.metadata.writer.LineageWriterDatasetFramework.writeLineage(LineageWriterDatasetFramework.java:180) ~[cdap-data-fabric-3.5.0.jar:na]
        at co.cask.cdap.data2.metadata.writer.LineageWriterDatasetFramework$BasicDatasetAccessRecorder.recordLineage(LineageWriterDatasetFramework.java:224) ~[cdap-data-fabric-3.5.0.jar:na]
        at co.cask.cdap.data2.dataset2.DefaultDatasetRuntimeContext.recordAccess(DefaultDatasetRuntimeContext.java:171) ~[cdap-data-fabric-3.5.0.jar:na]
        at co.cask.cdap.data2.dataset2.DefaultDatasetRuntimeContext.onMethodEntry(DefaultDatasetRuntimeContext.java:159) ~[cdap-data-fabric-3.5.0.jar:na]
        at co.cask.cdap.data2.dataset2.lib.table.AbstractTable.<init>(AbstractTable.java:73) ~[na:na]
        at co.cask.cdap.data2.dataset2.lib.table.BufferingTable.<init>(BufferingTable.java:137) ~[na:na]
        at co.cask.cdap.data2.dataset2.lib.table.hbase.HBaseTable.<init>(HBaseTable.java:98) ~[na:na]
        at co.cask.cdap.data2.dataset2.lib.table.hbase.HBaseTableDefinition.getDataset(HBaseTableDefinition.java:50) ~[cdap-data-fabric-3.5.0.jar:na]
        at co.cask.cdap.data2.dataset2.lib.table.hbase.HBaseTableDefinition.getDataset(HBaseTableDefinition.java:34) ~[cdap-data-fabric-3.5.0.jar:na]
        at co.cask.cdap.data2.transaction.queue.hbase.HBaseQueueDatasetModule$HBaseConsumerStateStoreDefinition.getDataset(HBaseQueueDatasetModule.java:89) ~[cdap-data-fabric-3.5.0.jar:na]
        at co.cask.cdap.data2.transaction.queue.hbase.HBaseQueueDatasetModule$HBaseConsumerStateStoreDefinition.getDataset(HBaseQueueDatasetModule.java:58) ~[cdap-data-fabric-3.5.0.jar:na]
        at co.cask.cdap.data2.datafabric.dataset.DatasetType.getDataset(DatasetType.java:73) ~[cdap-data-fabric-3.5.0.jar:na]
        at co.cask.cdap.data2.datafabric.dataset.RemoteDatasetFramework.getDataset(RemoteDatasetFramework.java:249) ~[cdap-data-fabric-3.5.0.jar:na]
        at co.cask.cdap.data2.dataset2.ForwardingDatasetFramework.getDataset(ForwardingDatasetFramework.java:160) ~[cdap-data-fabric-3.5.0.jar:na]
        at co.cask.cdap.data2.metadata.writer.LineageWriterDatasetFramework.access$101(LineageWriterDatasetFramework.java:50) ~[cdap-data-fabric-3.5.0.jar:na]
        at co.cask.cdap.data2.metadata.writer.LineageWriterDatasetFramework$2.call(LineageWriterDatasetFramework.java:166) ~[cdap-data-fabric-3.5.0.jar:na]
        at co.cask.cdap.data2.metadata.writer.LineageWriterDatasetFramework$2.call(LineageWriterDatasetFramework.java:163) ~[cdap-data-fabric-3.5.0.jar:na]
        at co.cask.cdap.data2.dataset2.DefaultDatasetRuntimeContext.execute(DefaultDatasetRuntimeContext.java:115) ~[cdap-data-fabric-3.5.0.jar:na]
        at co.cask.cdap.data2.metadata.writer.LineageWriterDatasetFramework.getDataset(LineageWriterDatasetFramework.java:162) ~[cdap-data-fabric-3.5.0.jar:na]
        ... 16 common frames omitted
```

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/DefaultDatasetRuntimeContext.java (hunk 25 above)
Comment (anew) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/DefaultDatasetRuntimeContext.java:145]: ok

**Comment by anew** [comment]
Changes look good to me. Please fix the problem pointed out by @anwar6953 before merging.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseQueueClientFactory.java [hunk 27]
```diff
@@ -84,6 +86,16 @@ public HBaseQueueClientFactory(CConfiguration cConf, Configuration hConf, HBaseT
   }
 
   @Override
+  public void initContext(Id.Run run) {
+    queueAdmin.initContext(run);
```
Comment (poornachandra) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseQueueClientFactory.java:90]: It is okay to set the run on singleton QueueAdmin since there will be only one run associated with a QueueAdmin, right?

**Comment by poornachandra** [comment]
Reviewed commit https://github.com/caskdata/cdap/pull/6179/commits/0f9f635b1c60909779746a24fa35ba2f16e1e7b7. Just one question.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseQueueClientFactory.java (hunk 27 above)
Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseQueueClientFactory.java:90]: The problem is not the queue admin, it is the `LineageWriterDatasetFramework` is not a singleton and it cannot be.

**Comment by chtyim** [comment]
Build passed: http://builds.cask.co/browse/CDAP-DUT4592-8
